docs: add `nextjs-instrumentation` docs (#305)
SigNoz/signoz.io: data/docs/instrumentation/nextjs.mdx

---
date: 2024-06-19
id: nextjs
title: Nextjs OpenTelemetry Instrumentation

> description:  Learn how to instrument your Nestjs application with OpenTelemetry and send telemetry data to SigNoz

Reviewer: ```suggestion
description:  Learn how to instrument your NextJS application with OpenTelemetry and send telemetry data to SigNoz
```

Author: ah, good catch!

---

> This document contains instructions on how to set up OpenTelemetry instrumentation in your Nextjs applications. OpenTelemetry, also known as OTel for short, is an open source observability framework that can help you generate and collect telemetry data - traces, metrics, and logs from your Nestjs application.

Reviewer: ```suggestion
This document contains instructions on how to set up OpenTelemetry instrumentation in your Nextjs applications. OpenTelemetry, also known as OTel for short, is an open source observability framework that can help you generate and collect telemetry data - traces, metrics, and logs from your NextJS application.
```

> ## Requirements

Reviewer: We should remove the requirements section for now since we are not sure if the version should be >=4.0.0 for NextJS as well.

Author: makes sense

- [Send traces via OTel Collector binary](#send-traces-via-otel-collector-binary) (recommended)

#### Send traces directly to SigNoz Cloud

**Step 1.** Install the OpenTelemetry packages

```bash
npm i @opentelemetry/sdk-node
npm i @opentelemetry/auto-instrumentations-node
npm i @opentelemetry/exporter-trace-otlp-http
npm i @opentelemetry/resources
npm i @opentelemetry/semantic-conventions
```

**Step 2.** Create `tracing.js` file<br></br>
You need to configure the endpoint for SigNoz cloud in this file.

```jsx
// tracing.js
'use strict'
const process = require('process');
const opentelemetry = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

// // Add otel logging when debugging
// const { diag, DiagConsoleLogger, DiagLogLevel } = require('@opentelemetry/api');
// diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

const exporterOptions = {

>   url: 'https://ingest.[data-region].signoz.cloud:443/v1/traces', // use your own data region 

Reviewer: ```suggestion
  url: 'https://ingest.{region}.signoz.cloud:443/v1/traces', // use your own data region 
```

Consistent with what we currently have for other instrumentation docs.

>   headers: { 'signoz-access-token': 'your-access-token' }, // Use if you are using SigNoz Cloud

Reviewer: ```suggestion
  headers: { 'signoz-access-token': 'SIGNOZ_INGESTION_KEY' }, // Use if you are using SigNoz Cloud
```

}

const traceExporter = new OTLPTraceExporter(exporterOptions);
const sdk = new opentelemetry.NodeSDK({
  traceExporter,
  instrumentations: [getNodeAutoInstrumentations()],
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'SigNoz-Nextjs-Example'
  })
});

// initialize the SDK and register with the OpenTelemetry API
// this enables the API to record telemetry
sdk.start()

// gracefully shut down the SDK on process exit
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('Tracing terminated'))
    .catch((error) => console.log('Error terminating tracing', error))
    .finally(() => process.exit(0));
});
```

- `<service_name>` : Name of the service you would like to monitor.

> - `signoz-access-token` : You can find your ingestion key from SigNoz cloud account details sent on your email.

Reviewer: ```suggestion
- `SIGNOZ_INGESTION_KEY` : You can find your ingestion key from SigNoz cloud account details sent on your email.
```

> - `http://localhost:4318/v1/traces` is the default url for sending your tracing data. We are assuming you have installed SigNoz on your `localhost`. Based on your environment, you can update it accordingly. It should be in the following format:

Reviewer: ```suggestion

```

This sentence is not valid for SigNoz cloud since the endpoint we use is `https://ingest.{region}.signoz.cloud:443/v1/traces`

`http://<IP of SigNoz backend>:4318/v1/traces`

> <Admonition>

Reviewer: Please remove this admonition as well, since its not needed for SigNoz cloud.

Remember to allow incoming requests to port 4318 of machine where SigNoz backend is hosted.
</Admonition>

Depending on the choice of your region for SigNoz cloud, the ingest endpoint will vary according to this table.

| Region | Endpoint |
| --- | --- |
| US |	ingest.us.signoz.cloud:443/v1/traces |
| IN |	ingest.in.signoz.cloud:443/v1/traces |
| EU | ingest.eu.signoz.cloud:443/v1/traces |

> **Step 3.** Once you're done configuring the exporter options, You can validate if your application is sending traces to SigNoz cloud [here](#validating-instrumentation-by-checking-for-traces).

Reviewer: We don't have the validation and troubleshooting section for NextJS right now so, we should remove step 3

Author: instead of removing the entire step, i think the only thing that needs to be removed is

_In case you encounter an issue where all applications do not get listed in the services section then please refer to the [troubleshooting section](#troubleshooting-your-installation)._

Reviewer: Cool. 
We don't have a section for validating instrumentation in the NextJS doc right now.

Please add the content available on this link - https://signoz.io/docs/instrumentation/nestjs/#validating-instrumentation-by-checking-for-traces after the conclusion section of the NextJS docs.

});
```

- `<service_name>` : Name of the service you would like to monitor.
- `SIGNOZ_INGESTION_KEY` : You can find your ingestion key from SigNoz cloud account details sent on your email.

Depending on the choice of your region for SigNoz cloud, the ingest endpoint will vary according to this table.

| Region | Endpoint |
| --- | --- |
| US |	ingest.us.signoz.cloud:443/v1/traces |
| IN |	ingest.in.signoz.cloud:443/v1/traces |
| EU | ingest.eu.signoz.cloud:443/v1/traces |

**Step 3.** Once you're done configuring the exporter options, You can validate if your application is sending traces to SigNoz cloud [here](#validating-instrumentation-by-checking-for-traces).

#### Send traces via OTel Collector binary

**Step 1.** Install OpenTelemetry Collector binary

OTel Collector binary helps to collect logs, hostmetrics, resource and infra attributes. It is recommended to install Otel Collector binary to collect and send traces to SigNoz cloud. You can correlate signals and have rich contextual data through this way.

You can find instructions to install OTel Collector binary [here](https://signoz.io/docs/tutorial/opentelemetry-binary-usage-in-virtual-machine/) in your VM. 

> While creating the `otel-config.yaml` during the installation fo the OTel Collector Binary, you need to enable CORS under the receivers section of the config file. This is needed so that you don't get 

Reviewer: ```suggestion
While creating the `config.yaml` during the installation fo the OTel Collector Binary, you need to enable CORS under the receivers section of the config file. This is needed so that you don't get 
```

> CORS error which can hinder sending your Traces to SigNoz Cloud. See the code snippet below to understand how you can enable CORS in your config file:

Reviewer: Please add the CORS snippet

**NOTE:** Make sure to restart your collector after making the config changes

**Step 2.** Install OpenTelemetry packages

```bash
npm i @opentelemetry/sdk-node
npm i @opentelemetry/auto-instrumentations-node
npm i @opentelemetry/exporter-trace-otlp-http
npm i @opentelemetry/resources
npm i @opentelemetry/semantic-conventions
```

> **Step 3.** In the `tracing.js` file, remove the headers from the exporterOptions since you're already using the signoz access token in the `otel-config.yml` file.

Reviewer: ```suggestion
**Step 3.** Create `tracing.js` file
```

Author: If we look carefully, the `tracing.js` flle is almost as same as the previous one except for exporter optons value. IMO, it is better to tell the user about this minute detail rather than asking to create the `tracing.js` file again.

Reviewer: The above section (Send Traces directly to Signoz Cloud) and this section (Send traces via OTel Collector binary) are independent sections. 
We should tell the user to create `tracing.js` file here as well since they might jump to following instructions in this section.

const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

// // Add otel logging when debugging
// const { diag, DiagConsoleLogger, DiagLogLevel } = require('@opentelemetry/api');
// diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

const exporterOptions = {
  url: 'https://ingest.[data-region].signoz.cloud:443/v1/traces', // use your own data region 
}

const traceExporter = new OTLPTraceExporter(exporterOptions);
const sdk = new opentelemetry.NodeSDK({
  traceExporter,
  instrumentations: [getNodeAutoInstrumentations()],
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'SigNoz-Nextjs-Example'
  })
});

// initialize the SDK and register with the OpenTelemetry API
// this enables the API to record telemetry
sdk.start()

// gracefully shut down the SDK on process exit
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('Tracing terminated'))
    .catch((error) => console.log('Error terminating tracing', error))
    .finally(() => process.exit(0));
});
```

> ```

Reviewer: ```suggestion

```
<img width="794" alt="Screenshot 2024-06-25 at 11 18 11 AM" src="https://github.com/SigNoz/signoz-web/assets/31571545/70038070-bc2d-400c-9c38-1acd442ad12c">

This was an additional ```

**Step 4.** Once we are done with the above configurations, you can run the collector service with the following command:
```bash
./otelcol-contrib --config ./config.yaml
```

**Step 5.** Start running your application and wait for few seconds to start receiving instrumentation data on the SigNoz Cloud. 

</TabItem>

<TabItem value="k8s" label="Kubernetes">

**Step 1.** Install OpenTelemetry Collector binary

OTel Collector binary helps to collect logs, hostmetrics, resource and infra attributes. It is recommended to install Otel Collector binary to collect and send traces to SigNoz cloud. You can correlate signals and have rich contextual data through this way.

You can find instructions to install OTel Collector binary [here](https://signoz.io/docs/tutorial/kubernetes-infra-metrics/) in your Kubernetes cluster. 

> While creating the `otel-config.yaml` during the installation fo the OTel Collector Binary, you need to enable CORS under the receivers section of the config file. This is needed so that you don't get 

Reviewer: ```suggestion

```
We don't have config file for Kubernetes

> CORS error which can hinder sending your Traces to SigNoz Cloud. See the code snippet below to understand how you can enable CORS in your config file:

Reviewer: ```suggestion

```

> **NOTE:** Make sure to restart your collector after making the config changes

Reviewer: ```suggestion

```

**Step 2.** Install OpenTelemetry packages

```bash
npm i @opentelemetry/sdk-node
npm i @opentelemetry/auto-instrumentations-node
npm i @opentelemetry/exporter-trace-otlp-http
npm i @opentelemetry/resources
npm i @opentelemetry/semantic-conventions
```

> **Step 3.** In the `tracing.js` file, remove the headers from the exporterOptions since you're already using the signoz access token in the `otel-config.yml` file.

Reviewer: ```suggestion
**Step 3.** Create `tracing.js` file
```

// diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);

const exporterOptions = {
  url: 'https://ingest.[data-region].signoz.cloud:443/v1/traces', // use your own data region 
}

const traceExporter = new OTLPTraceExporter(exporterOptions);
const sdk = new opentelemetry.NodeSDK({
  traceExporter,
  instrumentations: [getNodeAutoInstrumentations()],
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'SigNoz-Nextjs-Example'
  })
});

// initialize the SDK and register with the OpenTelemetry API
// this enables the API to record telemetry
sdk.start()

// gracefully shut down the SDK on process exit
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('Tracing terminated'))
    .catch((error) => console.log('Error terminating tracing', error))
    .finally(() => process.exit(0));
});
```
```

**Step 4.** Once we are done with the above configurations, you can run the collector service with the following command:
```bash
./otelcol-contrib --config ./config.yaml
```

**Step 5.** Start running your application and wait for few seconds to start receiving instrumentation data on the SigNoz Cloud. 

</TabItem>
</Tabs>

> ## Send traces to SigNoz Self-Host

Reviewer: Can we follow the same pattern as we did for Cloud here:

Step 1:

Step 2: 

Step 3:

...

  traceExporter,
  instrumentations: [getNodeAutoInstrumentations()],
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'SigNoz-Nextjs-Example'
  })
});

// initialize the SDK and register with the OpenTelemetry API
// this enables the API to record telemetry
sdk.start()

// gracefully shut down the SDK on process exit
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('Tracing terminated'))
    .catch((error) => console.log('Error terminating tracing', error))
    .finally(() => process.exit(0));
});
```

Again, `http://localhost:4318/v1/traces` is the default url for sending your tracing data. We are assuming you have installed SigNoz on your localhost. Based on your environment, you can update it accordingly. It should be in the following format:

`http://<IP of SigNoz backend>:4318/v1/traces`

**NOTE:** Remember to allow incoming requests to port 4318 of machine where SigNoz backend is hosted.

Once you're done with this, add your required changes in the `otel-config.yml` file which can be generally found [here](https://github.com/SigNoz/signoz/blob/develop/deploy/docker/clickhouse-setup/otel-collector-config.yaml).

> 

Reviewer: Add Validating traces section here just before conclusion section:

Checkout this comment: https://github.com/SigNoz/signoz-web/pull/305#discussion_r1651984968

Author: Should we try to add this before/after the conclusion section?

Reviewer: Before
